FIX tests

oliveralka opened the pull request:
There are still a few tests missing and I will try to fix them the following day(s).
 
```
     25 - StopWatch_test (Failed)
    FIXED 67 - BaseFeature_test (Failed)
    FIXED 71 - ConsensusFeature_test (Failed)
    FIXED 1203 - TOPP_IDMerger_6_out1 (Not Run)
    FIXED 1204 - TOPP_IDMerger_6_out2 (Failed)
    FIXED 1242 - TOPP_FeatureFinderIdentification_1 (Failed)
    FIXED 1243 - TOPP_FeatureFinderIdentification_1_out1 (Failed)
    FIXED 1244 - TOPP_FeatureFinderIdentification_3 (Failed)
    FIXED 1245 - TOPP_FeatureFinderIdentification_3_out1 (Failed)
    FIXED 1246 - TOPP_FeatureFinderIdentification_5 (Failed)
    FIXED 1247 - TOPP_FeatureFinderIdentification_5_out1 (Failed)
    FIXED 1591 - TOPP_IDFileConverter_27 (Failed)
    FIXED 1592 - TOPP_IDFileConverter_27_out1 (Failed)
    FIXED 1593 - TOPP_IDFileConverter_28 (Failed)
    FIXED 1594 - TOPP_IDFileConverter_28_out1 (Failed)
    FIXED 1595 - TOPP_IDFileConverter_29 (Failed)
    FIXED 1596 - TOPP_IDFileConverter_29_out1 (Failed)
    FIXED 1597 - TOPP_IDFileConverter_30 (Failed)
    FIXED 1598 - TOPP_IDFileConverter_30_out1 (Failed)
    FIXED 1609 - TOPP_IDFileConverter_34_out (Failed)
    FIXED 1698 - TOPP_MapAlignerIdentification_8 (Failed)
    FIXED 1699 - TOPP_MapAlignerIdentification_8_out1 (Failed)
    FIXED 2439 - UTILS_ProteomicsLFQ_1 (Failed)
    FIXED 2440 - UTILS_ProteomicsLFQ_1_out_1 (Failed)
    FIXED 2441 - UTILS_ProteomicsLFQ_1_out_2 (Failed)
    FIXED 2442 - UTILS_ProteomicsLFQ_1_out_3 (Failed)
    FIXED 2443 - UTILS_ProteomicsLFQ_1_out_4 (Failed)
    FIXED 2444 - UTILS_ProteomicsLFQ_2 (Failed)
    FIXED 2445 - UTILS_ProteomicsLFQ_2_out_1 (Failed)
    FIXED 2446 - UTILS_ProteomicsLFQ_2_out_2 (Failed)
    FIXED 2447 - UTILS_ProteomicsLFQ_2_out_3 (Failed)
    FIXED 2448 - UTILS_ProteomicsLFQ_2_out_4 (Failed)
    FIXED 2449 - UTILS_ProteomicsLFQ_3 (Failed)
    FIXED 2450 - UTILS_ProteomicsLFQ_3_out_1 (Failed)
    FIXED 2451 - UTILS_ProteomicsLFQ_3_out_2 (Failed)
    FIXED 2452 - UTILS_ProteomicsLFQ_3_out_3 (Failed)
    FIXED 2453 - UTILS_ProteomicsLFQ_3_out_4 (Failed)
    FIXED 2457 - UTILS_ProteomicsLFQ_5 (Failed)
    FIXED 2458 - UTILS_ProteomicsLFQ_5_out_1 (Failed)
    FIXED 2459 - UTILS_ProteomicsLFQ_5_out_2 (Failed)
    FIXED 2460 - UTILS_ProteomicsLFQ_5_out_3 (Failed)
    FIXED 2461 - UTILS_ProteomicsLFQ_5_out_4 (Failed)
    FIXED 2465 - UTILS_ProteomicsLFQ_7 (Failed)
    FIXED 2466 - UTILS_ProteomicsLFQ_7_out_1 (Failed)
    FIXED 2467 - UTILS_ProteomicsLFQ_7_out_2 (Failed)
    FIXED 2468 - UTILS_ProteomicsLFQ_7_out_3 (Failed)
    FIXED 2469 - UTILS_ProteomicsLFQ_7_out_4 (Failed) 
```

Issues: 

StopWatch_test
```
25: FAILED
25: Failed lines: 170 173 192 
1/1 Test #25: StopWatch_test ...................***Failed    0.44 sec

25: checking (bool stop()) ... 
25:  -  line 170:  TEST_EQUAL(s.getCPUTime() > t_wait / 2,true): got '0', expected '1'
25:  -  line 173:  TEST_EQUAL(s.getUserTime() > t_wait / 2,true): got '0', expected '1'
25:  -  line 192:  TEST_EQUAL(s_resume.getCPUTime() > (t_wait_more + t_wait) / 2,true): got '0', expected '1'
25: : failed
```

BaseFeature_test
```
67: FAILED
67: Failed lines: 183 
1/1 Test #67: BaseFeature_test .................***Failed    0.15 sec

checking (BaseFeature(BaseFeature &&feature)) ... 
 -  line 183:  TEST_EQUAL(noexcept(BaseFeature(std::declval<BaseFeature&&>())),true): got '0', expected '1'
: failed
```

ConsensusFeature_test
```
FAILED
Failed lines: 209 

 -  line 209:  TEST_EQUAL(noexcept(ConsensusFeature(std::declval<ConsensusFeature&&>())),true): got '0', expected '1'
: failed
```

TOPP_FeatureFinderIdentification
```
1243: Error: File not found (the file 'FeatureFinderIdentification_1.tmp' could not be found)
1243: Cannot read input file given from parameter '-in1'!
2/2 Test #1243: TOPP_FeatureFinderIdentification_1_out1 ...***Failed    0.05 sec

1245: Error: File not found (the file 'FeatureFinderIdentification_3.tmp' could not be found)
1245: Cannot read input file given from parameter '-in1'!
2/2 Test #1245: TOPP_FeatureFinderIdentification_3_out1 ...***Failed    0.07 sec

1247: Error: File not found (the file 'FeatureFinderIdentification_5.tmp' could not be found)
1247: Cannot read input file given from parameter '-in1'!
2/2 Test #1247: TOPP_FeatureFinderIdentification_5_out1 ...***Failed    0.06 sec
```

MapAlignerIdentification
```
1698: Test command: /Users/alka/Documents/work/software/openms/openms_build/bin/MapAlignerIdentification "-test" "-in" "/Users/alka/Documents/work/software/openms/OpenMS/src/tests/topp/MapAlignerIdentification_8_input1.oms" "-trafo_out" "MapAlignerIdentification_8_output1.tmp" "-out" "MapAlignerIdentification_8_output2.tmp" "-reference:file" "/Users/alka/Documents/work/software/openms/OpenMS/src/tests/topp/MapAlignerIdentification_8_input2.oms" "-store_original_rt"
1698: Test timeout computed to be: 1500
1698: Error: Unexpected internal error (Could not extract retention time information from the reference file)
1/2 Test #1698: TOPP_MapAlignerIdentification_8 ........***Failed    0.24 sec
test 1699
    Start 1699: TOPP_MapAlignerIdentification_8_out1
```

ProtemicsLFQ
```
2439: Error: Unexpected internal error (no match for reference)
1/5 Test #2439: UTILS_ProteomicsLFQ_1 ............***Failed    0.99 sec
test 2440
```
Some of them seem to be related to it `IdentificationDataConverter::exportFeatureIDs()`

jpfeuffer commented:
StopWatch is usually a spurious error.

timosachsenberg commented:
-  line 183:  TEST_EQUAL(noexcept(BaseFeature(std::declval<BaseFeature&&>())),true): got '0', expected '1' is a tougher one. THis means that a move constructor that does not throw does not exist. Usually means that all members need to have one of those and I guess the ID structure members don't

oliveralka commented:
It seems for FFIdent test 1 and 3 have changed a bit. 

In the case of test 1 one ProteinHit is missing

```
<ProteinHit id="PH_5" accession="tr|A9GV08|A9GV08_SORC5" score="0" sequence="">
	<UserParam type="float" name="OMSSA_score" value="0.0"/>
</ProteinHit>
```

Running the test via command line led to:

```
Writing final results...
While storing '': Omitting peptide identification because of missing ProteinIdentification with identifier '' while writing '/Users/alka/Desktop/idfile/featurefinderidentification/FeatureFinderIdentification_1_output_clion.tmp.featureXML'!
FeatureFinderIdentification took 2.86 s (wall), 2.69 s (CPU), 0.08 s (system), 2.61 s (user); Peak Memory Usage: 100 MB.
<While storing '': Omitting peptide identification because of missing ProteinIdentification with identifier '' while writing '/Users/alka/Desktop/idfile/featurefinderidentification/FeatureFinderIdentification_1_output_clion.tmp.featureXML'!> occurred 24 times
```
Is that alright?

jpfeuffer commented:
Does not sound very good.

timosachsenberg commented on src/openms/source/KERNEL/FeatureMap.cpp:
```diff
@@ -126,14 +126,34 @@ namespace OpenMS
+      data_processing_(source.data_processing_),
+      id_data_() // updated below
+      {
+        // copy ID data and update references in features:
```
shouldn't the move constructor work without updating the references?

timosachsenberg replied:
e.g., does the default move work?

oliveralka replied:
Yes, seems to work.

timosachsenberg commented on src/openms/source/TRANSFORMATIONS/FEATUREFINDER/FeatureFinderIdentificationAlgorithm.cpp:
```diff
@@ -504,8 +504,16 @@ namespace OpenMS
     {
-      Size pos_hash = target_id.find('#', pos_slash + 2);
-      charge = target_id.substr(pos_slash + 1, pos_hash).toInt();
+      if (target_id.find("#") != std::string::npos) // "PEP:XXXXX/3#1
+      {
+        Size pos_hash = target_id.find('#', pos_slash + 2);
+        size_t charge_size = pos_hash - pos_slash - 1;
+        charge = target_id.substr(pos_slash + 1, charge_size).toInt();
+      }
+      else // "PEP:XXXXX/2
+      {
+        charge = target_id.substr(pos_slash + 1, target_id.size()).toInt();
+      }
```
not very efficient as it parses the string multile times

oliveralka replied:
Tried to clean it up.

oliveralka commented:
There are mulitples issues with the ProteomicsLFQ tests.

**First issue:** 
ProteomicsLFQ fails to run, due to empty PeptideIdentifications, which are linked to a Feature. 
https://github.com/OpenMS/OpenMS/blob/fix/hendrik_idfile_integration/src/openms/source/TRANSFORMATIONS/FEATUREFINDER/FeatureFinderIdentificationAlgorithm.cpp#L557

StackTrace:
```
OpenMS::MzTab::peptideSectionRowFromConsensusFeature_(OpenMS::ConsensusFeature const&, OpenMS::ConsensusMap const&, std::__1::vector<OpenMS::String, std::__1::allocator<OpenMS::String> > const&, unsigned long, std::__1::set<OpenMS::String, std::__1::less<OpenMS::String>, std::__1::allocator<OpenMS::String> > const&, std::__1::set<OpenMS::String, std::__1::less<OpenMS::String>, std::__1::allocator<OpenMS::String> > const&, std::__1::map<OpenMS::String, unsigned long, std::__1::less<OpenMS::String>, std::__1::allocator<std::__1::pair<OpenMS::String const, unsigned long> > > const&, std::__1::map<std::__1::pair<unsigned long, unsigned long>, unsigned long, std::__1::less<std::__1::pair<unsigned long, unsigned long> >, std::__1::allocator<std::__1::pair<std::__1::pair<unsigned long, unsigned long> const, unsigned long> > > const&, std::__1::map<std::__1::pair<OpenMS::String, unsigned int>, unsigned int, std::__1::less<std::__1::pair<OpenMS::String, unsigned int> >, std::__1::allocator<std::__1::pair<std::__1::pair<OpenMS::String, unsigned int> const, unsigned int> > > const&, std::__1::vector<OpenMS::String, std::__1::allocator<OpenMS::String> > const&, bool) MzTab.cpp:1843
OpenMS::MzTab::CMMzTabStream::nextPEPRow(OpenMS::MzTabPeptideSectionRow&) MzTab.cpp:3747
OpenMS::MzTab::exportConsensusMapToMzTab(OpenMS::ConsensusMap const&, OpenMS::String const&, bool, bool, bool, bool, bool, OpenMS::String const&) MzTab.cpp:3822
ProteomicsLFQ::main_(int, char const**) ProteomicsLFQ.cpp:1942
OpenMS::TOPPBase::main(int, char const**) TOPPBase.cpp:444
main ProteomicsLFQ.cpp:1999
```

Commenting the lines introducing the "dummy" PeptideIdentification solves the issue. 

@hendrikweisser: Do you think these "dummys" are still needed here?

**Second issue - ConsensusXML Output:** 
@jpfeuffer: Could you please have a look at the diff in the last commit? Does that look alright, or what kind of changes would be necessary? 

Changed:
`charges="1, 2, 3"`

Added:
`<UserParam type="string" name="target_decoy" value="target"/>`

Removed:
`<UserParam type="int" name="fraction_group" value="1"/>`
`<UserParam type="int" name="fraction" value="1"/>`
Now only available in `mapList`

Moved:
`<UserParam type="string" name="FFId_category" value="internal"/>` moved inside PeptideHit

MzTab:
Added `opt_global_FFId_category`

**Third issue:**
ProteomicsLFQ_3 
`2449: Error: Unexpected internal error (input file must have a name)`
Have not yet investigated further.

jpfeuffer commented on src/openms/source/TRANSFORMATIONS/FEATUREFINDER/FeatureFinderIdentificationAlgorithm.cpp:
```diff
@@ -554,6 +552,19 @@ namespace OpenMS
       feature.setMetaValue("label", target_id);
 
       // TODO: Why is the dummy data set here?
```
@hendrikweisser Can you comment on the necessity of the dummy ID here?

hendrikweisser replied:
Good point. I think that should be removed (left-over from the previous version that used `IdentificationData` internally, but converted to `PeptideIdentification` for output).

jpfeuffer commented on src/tests/topp/ProteomicsLFQ_1_out.consensusXML:
```diff
@@ -205,624 +205,648 @@
 		<ProteinIdentification score_type="FDR" higher_score_better="false" significance_threshold="0">
 			<ProteinHit id="PH_25" accession="P02769|ALBU_BOVIN" score="0" sequence="">
 				<UserParam type="float" name="OMSSA_score" value="0.0"/>
+				<UserParam type="string" name="target_decoy" value="target"/>
```
Seems like TD information goes missing.

oliveralka commented:
@jpfeuffer: Could you please have another look at the last commit to check the files for ProteomicsLFQ3 (https://github.com/hendrikweisser/OpenMS/pull/14/commits/d398538e0ef43e3ce8bf0572f5a8cc830d0548b4)

The problem here was that the FilePath in DocumentIdentifier in the FeatureMap, which is needed in the FeatureFinderIdentificationAlgorithm was not set previously and was reset in the `calculateSeeds_` function. Now they are reset in the "if clause".

jpfeuffer commented on src/tests/topp/ProteomicsLFQ_3_out.csv:
```diff
@@ -1,88 +1,69 @@
-P02769|ALBU_BOVIN,C(Carbamidomethyl)C(Carbamidomethyl)TESLVNR,2,NA,0,L,2,2,3,1,1.103722e07,"BSA2_F1.mzML"
-P02769|ALBU_BOVIN,C(Carbamidomethyl)C(Carbamidomethyl)TESLVNR,2,NA,0,L,3,3,5,1,1.794934e06,"BSA3_F1.mzML"
-P02769|ALBU_BOVIN,CC(Carbamidomethyl)TESLVNR,2,NA,0,L,3,3,5,1,2.029336e05,"BSA3_F1.mzML"
-P02769|ALBU_BOVIN,DDPHAC(Carbamidomethyl)YSTVFDK,3,NA,0,L,2,2,4,2,1.087751e07,"BSA2_F2.mzML"
```
Hmm this was completely lost.

oliveralka replied:
Any idea why? Should not be related to the setting PATH issue, or?

oliveralka replied:
should be resolved with Hendriks fix.

jpfeuffer commented on src/tests/topp/FeatureFinderIdentification_1_output.featureXML:
```diff
@@ -1,549 +1,454 @@
-				<UserParam type="float" name="OMSSA_score" value="0.0"/>
-			</ProteinHit>
-			<ProteinHit id="PH_14" accession="P02769|ALBU_BOVIN" score="0" sequence="">
+			<ProteinHit id="PH_13" accession="tr|A9GN44|A9GN44_SORC5" score="0" sequence="">
```
I think we should focus on the FFID tests first. Not even here, the numbers are the same. Was something changed in the ProteinHit Comparator? They are sorted completely in the opposite direction. Pretty annoying for comparison.
EDIT: As discussed with @oliveralka we probably need to sort at the end of the export or before writing to file. (a general strategy for when to assume sortedness would be nice as well).

timosachsenberg replied:
agree. first these issues

oliveralka replied:
@hendrikweisser Do you have any idea where that could go missing?

oliveralka replied:
@jpfeuffer I am not sure about changes in the ProteinHit Comparator. Could you give me a pointer where to look?

The problem is that normal `ProteinIdentification::sort()` does not do anything, since it will sort by score (0).

jpfeuffer replied:
It should also sort by accession

oliveralka replied:
See comment below.

oliveralka commented:
As jpfeuffer already stated in the meeting, it seems that the `noexcept` statement for BaseFeature and ConsensusFeature is not true anymore. 

Trying to enforce noexcept:
`BaseFeature(BaseFeature&& feature) noexcept = default;`
leads to:
`error: exception specification of explicitly defaulted move constructor does not match the calculated one`

This is probably due to the introduction of the identification data elements, which are not `noexcept`.
    
 Since noexcept was used to:
```
// Ensure that BaseFeature has a no-except move constructor (otherwise
// std::vector is inefficient and will copy instead of move).
```

I am not sure how to proceed here.

jpfeuffer commented:
IIRC there are already some commented-out tests where VS complains. Just comment-out BaseFeature tests with a comment as well.

timosachsenberg commented:
I think we could do what Hendrik proposed. Implement the move constructor for that data structure manually and mark it noexcept.
So basically do .swap for vector and set and just assign the others.
Note that in the case of an exception (like out of memory) the program will terminate with an error, but usually one can not really recover from this type of error anyways.

hendrikweisser commented:
> Implement the move constructor for that data structure manually and mark it noexcept.

We should do that at the "lowest level" possible. For example, if we add a noexcept move c'tor to `IteratorWrapper` (the base class for the references in `IdentificationData`), perhaps the other ones can be derived automatically.

hendrikweisser commented on src/openms/source/ANALYSIS/ID/FalseDiscoveryRate.cpp:
```diff
@@ -582,8 +582,10 @@ namespace OpenMS
       vector<IdentificationData::ObservationMatchRef> best_matches =
           id_data.getBestMatchPerObservation(score_ref);
-      for (auto match_ref : best_matches)
+      for (const auto& match_ref : best_matches)
```
When dealing with ID references, I generally just use `auto` because they are iterators, so the overhead of making a copy should be the same as that of taking a reference via "&".

hendrikweisser commented on src/openms/source/ANALYSIS/ID/FalseDiscoveryRate.cpp:
```diff
@@ -582,8 +582,10 @@ namespace OpenMS
-      for (auto match_ref : best_matches)
+      for (const auto& match_ref : best_matches)
       {
+
```
What's with the empty lines?

hendrikweisser commented on src/openms/source/KERNEL/BaseFeature.cpp:
```diff
@@ -253,13 +253,13 @@ namespace OpenMS
     set<IdentificationData::ObservationMatchRef> matches;
     matches.swap(id_matches_);
-    for (auto item : matches)
+    for (const auto& item : matches)
```
See comment regarding "auto" above.

hendrikweisser commented on src/openms/source/KERNEL/FeatureMap.cpp:
```diff
@@ -126,14 +126,16 @@ namespace OpenMS
-    for (Feature& feature : *this)
     {
-      feature.updateAllIDReferences(trans);
+      // copy ID data and update references in features:
```
Did only the indentation change here? I think it was correct before (aligned with `FeatureMap::FeatureMap(...)`.

timosachsenberg commented:
Agree. Problems are usually caused by std::set and std::map (and friends) though

hendrikweisser commented on src/openms/source/METADATA/ID/IdentificationDataConverter.cpp:
```diff
@@ -1058,14 +1058,14 @@ namespace OpenMS
         for (const String& key : meta_keys)
         {
-          if (key.hasPrefix("IDConverter_trace_"))
+          if (key.hasPrefix("IDConverter_trace_")) // TODO: What is happening here (comment)
```
There's a comment in `handleFeatureExport_`, where the meta value is set:
```
      // store trace of feature indexes so we can map the converted ID back;
      // key needs to be unique in case the same ID matches multiple features:
```
Does that make it clear?
If not: Before the ID conversion, in every ID that is attached to a feature (or subfeature!) I store a "trace" (path through the feature-subfeature hierarchy) to that feature. Then after the IDs are converted, I look for these traces and use them to attach converted IDs to features again.

timosachsenberg replied:
after putting a bit of thought into the code we understood what it does. Would just be great to add a more extensive comment here so others don't take like 15 minutes to figure it out

oliveralka replied:
Thanks - Hendrik! 
I will add a comment.

hendrikweisser commented on src/openms/source/METADATA/ID/IdentificationDataConverter.cpp:
```diff
@@ -1102,10 +1102,17 @@ namespace OpenMS
+    if (clear_original)
+    {
+      features.getIdentificationData().clear();
+      for (auto& feat : features)
```
Thanks, well spotted!

hendrikweisser commented on src/openms/source/TRANSFORMATIONS/FEATUREFINDER/FeatureFinderIdentificationAlgorithm.cpp:
```diff
@@ -495,19 +494,27 @@ namespace OpenMS
   {
     String target_id = feature.getMetaValue("PeptideRef");
-    Size pos_slash = target_id.rfind('/');
+
```
Why did you make changes in this function? Was there a bug? I'm asking because to me it looked simpler/cleaner before.

jpfeuffer replied:
I think there was a bug. @oliveralka knows more.

oliveralka replied:
PEP:XXXXX/2  did lead to an error since no was hash could be found in the string.
That is why I added two options (if/else). 
This could probably be handled much cleaner with a more complicated parsing as Timo mentioned once.

hendrikweisser replied:
> PEP:XXXXX/2 did lead to an error since no was hash could be found in the string.

That can't be true. If there's no hash, the `find` call returns `npos` and the `substr` call reads to the end of the string. So you get the correct result.

oliveralka replied:
Seems it was the other way round, sorry I will double-check:

```
ProteomicsLFQ_tests:
Error: Unexpected internal error (Prefix of string '3#1' successfully converted to an integer value. Additional characters found at position 2)
```

Maybe I missed something and you wanted to extract it in that fashion?

hendrikweisser replied:
Okay, there was indeed a bug, but for reference strings *with* a hash, e.g. "PEP:XXXXX/2#1". The reason was the second argument in the `substr` call: That should be a character count, *not* the end position of the substring. I've fixed it.

oliveralka replied:
thank you!

hendrikweisser commented:
@timosachsenberg:
> Problems are usually caused by std::set and std::map (and friends) though

Okay. It could well be `std::set` in `BaseFeature`, then.

oliveralka commented:
All tests are fixed (in theory), but the issue with the missing ProteinHit persists - I will look further into it.

oliveralka commented:
In terms of the "missing" ProteinHit in the test: FeatureFinderIdentification_1 

https://github.com/OpenMS/OpenMS/blob/fix/hendrik_idfile_integration/src/tests/topp/FeatureFinderIdentification_1_input.idXML#L25
```
<ProteinHit id="PH_5" accession="tr|A9GV08|A9GV08_SORC5" score="0" sequence="">
	<UserParam type="float" name="OMSSA_score" value="0.0"/>
</ProteinHit>
```
PH_5 does not have any reference to a PeptideIdentification (already in the input file).

It is filtered out by the` IDFilter::keepBestMatchPerObservation` by calling `id_data.cleanup();` in the end. 
https://github.com/OpenMS/OpenMS/blob/fix/hendrik_idfile_integration/src/openms/source/TRANSFORMATIONS/FEATUREFINDER/FeatureFinderIdentificationAlgorithm.cpp#L282

https://github.com/OpenMS/OpenMS/blob/fix/hendrik_idfile_integration/src/openms/source/FILTERING/ID/IDFilter.cpp#L849

As far as I have understood because of 
```
"require_identified_sequence = true" (default)
@param require_identified_sequence Remove parent sequences (proteins/RNAs) that aren't referenced by identified peptides/oligonucleotides?
```
https://github.com/OpenMS/OpenMS/blob/fix/hendrik_idfile_integration/src/openms/include/OpenMS/METADATA/ID/IdentificationData.h#L532

hendrikweisser commented:
> PH_5 does not have any reference to a PeptideIdentification (already in the input file).

Nice sleuthing, @oliveralka! Is this the reason for all the protein hits that get removed?
If so:
- Should they be removed or not?
- If they should be kept, does setting the `require_identified_sequence` flag in the `cleanup` function to `false` solve the problem? (The flag probably needs to be exposed in `keepBestMatchPerObservation`...)

oliveralka commented:
It solves the issue with the FeatureFinderIdentification regarding the missing ProteinHit (test 1,3,5), but it does not seem to have anything to do with the ProteomicsLFQ changes. 

@jpfeuffer What do you think it terms of Hendriks statements above, should ProteinHits which do not have a PeptideIdentification attached be filtered out? Then the current test for FFI would be correct.

jpfeuffer commented:
Remove

hendrikweisser commented:
I'm just thinking about whether it's really ideal that FFId filters the ID data. For quantification we only want to consider the best hit (`ObservationMatch`) per spectrum. Currently that's accomplished by removing all other hits via `IDFilter::keepBestMatchPerObservation`, then iterating over the remaining hits. It may be better to iterate over the result of `IdentificationData::getBestMatchPerObservation` instead and keep the ID data unchanged.

As far as I can see, the code would only need to be changed in two places (in `FFIdAlgo::run`, for the internal and external IDs) and then it should just work. But I haven't tried it yet.

Any thoughts?

jpfeuffer commented:
I like keeping the IDs around as long as possible. Linking for example might give even more information to resolve ambiguities in features (not only spectra).

(EDIT: although the results/scores in FFID are pretty much tailored to the best match only, right?)

oliveralka commented:
There is another issue regarding the `seeds` and the `IDFilter::keepBestMatchPerObservation`. 

Constructing the assay library leads to the following output:
```
OpenMS develop: Will analyse 51 peptides with a total of 102 transitions 
Current branch: Will analyse 15 peptides with a total of 30 transitions
```
As far as I have seen the `IDFilter::keepBestMatchPerObservation` filters by score

```
    vector<IdentificationData::ObservationMatchRef> best_matches =
      id_data.getBestMatchPerObservation(score_ref);
```

I found that when converting the seeds `FeatureFinderIdentificationAlgorithm::convertSeeds` no score information is provided, and all seeds are filtered out in the IDFilter. 

Adding "dummy" scores to the seeds in `FeatureFinderIdentificationAlgorithm::convertSeeds` the rescues the assay library:
```
ID::ScoreTypeRef score_ref = id_data.pickScoreType(id_data.getObservationMatches());
id_data.addScore(match_ref, score_ref, 0.0);
```

`Branch_with_changes: Will analyse 51 peptides with a total of 102 transitions `

Unfortunately it leads to an error down the line, which I have to investigate further:
```
Error: Unexpected internal error (Header and content differs in columns. Please report this bug to the OpenMS developers.)
Error occurred in line 166 of file /Users/alka/Documents/work/software/openms/OpenMS/src/openms/include/OpenMS/FORMAT/MzTabFile.h (in function: void OpenMS::MzTabFile::generateMzTabSection_(const std::vector<SectionRow> &, const std::vector<String> &, const OpenMS::MzTabMetaData &, OpenMS::StringList &, size_t) const [SectionRow = OpenMS::MzTabPeptideSectionRow]) !
```

Edit: This has nothing to do with the `id_data.cleanup()` part.

timosachsenberg commented:
Why are these filtered out? Is this a bug in IDFilter? @hendrikweisser

hendrikweisser commented:
> Is this a bug in IDFilter?

Looks like it. I guess it's safe to say that for `keepBestMatchPerObservation`, nothing should be done if there is only one match to begin with. I'll put a fix in.
Thanks for pinpointing the problem, @oliveralka.

hendrikweisser commented:
I think I've fixed the IDFilter problem by changing `IdentificationData::getBestMatchPerObservation` (which is used by `IDFilter::keepBestMatchPerObservation`). @oliveralka, can you give it a try?

oliveralka commented:
@hendrikweisser Thanks - I will give it a try! 

Edit: Nevermind, will merge your fixes :)

oliveralka commented:
Seems to work - it looks like the PLFQ3 test is now similar to the dev version.

oliveralka commented:
It seems there is an inconsistency. The corrected tests fail sometimes for ProteomcisLFQ3 - I am not sure why - will try to reproduce. 

Edit: It seems there are two possible identifications for a specific feature. 

line 486 (consensusXML):
Current test (valid most of the time)
```
	<UnassignedPeptideIdentification identification_run_ref="PI_1" score_type="Posterior Error Probability" higher_score_better="false" significance_threshold="0" MZ="379.715393066406" RT="1886.87810354796" spectrum_reference="spectrum=2624" >
		<PeptideHit score="0" sequence="YIC(Carbamidomethyl)DNQDTISSK" charge="2" aa_before="K" aa_after="L" protein_refs="PH_25">
			<UserParam type="float" name="OMSSA_score" value="3.85390695500564e-05"/>
			<UserParam type="string" name="target_decoy" value="target"/>
			<UserParam type="string" name="FFId_category" value="internal"/>
		</PeptideHit>
		<UserParam type="string" name="feature_id" value="not mapped"/>
		<UserParam type="int" name="map_index" value="0"/>

```

Test fail due to different identification of the same feature: 
```
	<UnassignedPeptideIdentification identification_run_ref="PI_1" score_type="Posterior Error Probability" higher_score_better="false" significance_threshold="0" MZ="379.715393066406" RT="1886.87810354796" spectrum_reference="spectrum=2624" >
		<PeptideHit score="0" sequence="GAC(Carbamidomethyl)LLPK" charge="2" aa_before="K" aa_after="I" protein_refs="PH_25">
			<UserParam type="float" name="OMSSA_score" value="8.696044183457621e-03"/>
			<UserParam type="string" name="target_decoy" value="target"/>
			<UserParam type="string" name="FFId_category" value="internal"/>
		</PeptideHit>
		<UserParam type="string" name="feature_id" value="not mapped"/>
		<UserParam type="int" name="map_index" value="0"/>
```

Edit: I am still not sure what is going on. Maybe it has something to do with the `tranferIDsBetweenSameFractions_`:

BSA1_F1 (idXML):
```
<PeptideIdentification score_type="Posterior Error Probability" higher_score_better="false" significance_threshold="0" MZ="722.32470703125" RT="1804.15795898438" >
    <PeptideHit score="0" sequence="YIC(Carbamidomethyl)DNQDTISSK" charge="2" aa_before="K" aa_after="L" protein_refs="PH_0" >
        <UserParam type="float" name="OMSSA_score" value="3.85390695500564e-05"/>
        <UserParam type="string" name="target_decoy" value="target"/>
    </PeptideHit>
```

BSA3_F1 (idXML):
```
<PeptideIdentification score_type="Posterior Error Probability" higher_score_better="false" significance_threshold="0" MZ="379.715393066406" RT="1887.15539550781" >
    <PeptideHit score="0" sequence="GAC(Carbamidomethyl)LLPK" charge="2" aa_before="K" aa_after="I" protein_refs="PH_1" >
        <UserParam type="float" name="OMSSA_score" value="0.00869604418345762"/>
        <UserParam type="string" name="target_decoy" value="target"/>
</PeptideHit>
```

The Ids above are the ones that are exchanged. In around 60% of the cases `YIC(Carbamidomethyl)DNQDTISSK` is used. 

mztab (60%):
```
PSM    YICDNQDTISSK    79    P02769|ALBU_BOVIN    1    18Protein_SoCe_Tr_detergents_trace_target_decoy    null    [, , OMSSA, 2.1.4]    0.0    3-UNIMOD:4    1886.878103547961928    2    379.715393066406023    722.324660331071073    ms_run[1]:spectrum=2624    K    L    null    null    not mapped    0    spectrum=2624    YIC(Carbamidomethyl)DNQDTISSK    internal    3.85390695500564e-05    0
```

mztab (40%):
```
PSM    GACLLPK    79    P02769|ALBU_BOVIN    1    18Protein_SoCe_Tr_detergents_trace_target_decoy    null    [, , OMSSA, 2.1.4]    0.0    3-UNIMOD:4    1886.878103547961928    2    379.715393066406023    379.715100772820961    ms_run[1]:spectrum=2624    K    I    null    null    not mapped    0    spectrum=2624    GAC(Carbamidomethyl)LLPK    internal    8.696044183457621e-03    0
```
@jpfeuffer: Do you have an Idea where this behaviour could be introduced? 
I am also wondering about the high difference between the expected and calculated mass to charge in the case of YICDNQDTISSK?

jpfeuffer commented:
Hmm but they are not really features since they are unmapped. However, the current PSM seems to be wrong indeed.
Which spectrum references do they have? Unfortunately this info is not present in those old idXMLs and only gets added in the tool (via SpectrumLookup).

hendrikweisser commented:
Yes, the "current" PSM (in "ProteomicsLFQ_3_out.consensusXML", "YIC(Carbamidomethyl)DNQDTISSK") is definitely wrong; the `PeptideHit` doesn't match the `(Unassigned)PeptideIdentification` - e.g. wrong m/z. The question is, how would this `PeptideHit` ever end up in the wrong `PeptideIdentification`?

jpfeuffer commented:
The problem is that when transferring IDs from other maps, you may have the same spectrum_ref in two Peptide IDs of the same "idXML" (`vec<PepID>`).

https://github.com/OpenMS/OpenMS/blob/develop/src/utils/ProteomicsLFQ.cpp#L1073

IdentificationData cannot handle that and corrupts old data for this spectrum_ref during importing (after encountering the second ID with the same spectrum_ref). Why does it not fail or warn?

We probably have to annotate the file_origin when combining internal and transferred (not external!) IDs. Or pass a third type of IDs to FFID (could also be used for seeds).

jpfeuffer commented:
There is also something wrong with the proteins, I tried another test file (intermediate PLFQ idXML or also just BSA1_F1.idXML) in the IDConv_test for the round-trip conversion, and it fails when comparing proteins before and after. Strangely, the pepID comparison passes although it includes the spectrum_ref anomaly. Maybe it is not testing enough?

timosachsenberg commented:
@hendrikweisser could you take a look at that issue?
E.g., also add some of round-trip conversions to the tests?

jpfeuffer commented:
There is one roundtrip test already. I'll provide another file. As I said, I think the comparison of peptides maybe needs to be more detailed. I think it does not fail for duplicate spec_refs.
The following should be the actual output of the failing step in PLFQ
[debug_fraction_1_IDs_after_transfer.idXML.gz](https://github.com/hendrikweisser/OpenMS/files/6451200/debug_fraction_1_IDs_after_transfer.idXML.gz)

jpfeuffer commented:
So, this file actually does not include a duplicate spectrum_reference, which maybe is the cause for the 50:50 occurrence of this bug. Maybe something in the transfer or the spectrum_lookup is nondeterministic?

Commit:
Bit of cleanup and some debug comments

Commit:
add the correct test?

timosachsenberg committed:
renamed operator+= and more error messages

Commit:
Merge remote-tracking branch 'upstream/develop' into fix/hendrik_idfile_integration

Commit:
additional debug output in exception

Commit:
Merge remote-tracking branch 'hendrikweisser/idfile-integration' into fix/hendrik_idfile_integration

timosachsenberg commented:
throwing the exceptions seem to expose some problems:
The following tests FAILED:
        110 - IdentificationDataConverter_test (Failed)
        197 - OMSFile_test (SEGFAULT)
        1612 - TOPP_IDFileConverter_34_1 (Failed)
        1613 - TOPP_IDFileConverter_34_2 (Failed)
        1614 - TOPP_IDFileConverter_34_out (Failed)
        2421 - UTILS_ProteomicsLFQ_2 (Failed)
        2422 - UTILS_ProteomicsLFQ_2_out_1 (Failed)
        2424 - UTILS_ProteomicsLFQ_2_out_3 (Failed)
        2429 - UTILS_ProteomicsLFQ_3_out_3 (Failed)
        2430 - UTILS_ProteomicsLFQ_3_out_4 (Failed)

timosachsenberg commented:
Erros are mainly:
Error: Caught unexpected exception of type 'InvalidValue' thrown in line 88 of file '/home/sachsenb/OMS/OpenMS/src/openms/include/OpenMS/METADATA/ID/ParentSequence.h' in function 'OpenMS::IdentificationDataInternal::ParentSequence& OpenMS::IdentificationDataInternal::ParentSequence::merge(const OpenMS::IdentificationDataInternal::ParentSequence&)' - Message: the value '' was used but is not valid; Trying to overwrite ParentSequence sequence 'HIJKLMN' with conflicting value.

timosachsenberg commented:
```
Progress of 'Storing...':

  Progress of 'converting protein identification runs':

Catchpoint 1 (exception thrown), 0x00007ffff3f45d1d in __cxa_throw () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
(gdb) bt
#0  0x00007ffff3f45d1d in __cxa_throw () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#1  0x00007ffff5d79fb5 in OpenMS::StringUtils::toInt (this_s=...) at /home/sachsenb/OMS/OpenMS/src/openms/include/OpenMS/DATASTRUCTURES/StringUtils.h:810
#2  0x00007ffff5d74e1a in OpenMS::String::toInt (this=0x55555596cd88) at /home/sachsenb/OMS/OpenMS/src/openms/source/DATASTRUCTURES/String.cpp:293
#3  0x00007ffff5e1c968 in OpenMS::ProteinIdentification::SearchParameters::getChargeRange (this=0x55555596cd20) at /home/sachsenb/OMS/OpenMS/src/openms/source/METADATA/ProteinIdentification.cpp:208
#4  0x00007ffff5ec8d79 in OpenMS::IdentificationDataConverter::importDBSearchParameters_ (pisp=..., id_data=...) at /home/sachsenb/OMS/OpenMS/src/openms/source/METADATA/ID/IdentificationDataConverter.cpp:888
#5  0x00007ffff5ec1956 in OpenMS::IdentificationDataConverter::importIDs (id_data=..., proteins=std::vector of length 2, capacity 2 = {...}, peptides=std::vector of length 3, capacity 4 = {...})
    at /home/sachsenb/OMS/OpenMS/src/openms/source/METADATA/ID/IdentificationDataConverter.cpp:74
#6  0x00005555555921f8 in TOPPIDFileConverter::main_ (this=0x7fffffffd9a0) at /home/sachsenb/OMS/OpenMS/src/topp/IDFileConverter.cpp:807
#7  0x00007ffff70a7ae4 in OpenMS::TOPPBase::main (this=0x7fffffffd9a0, argc=6, argv=0x7fffffffe098) at /home/sachsenb/OMS/OpenMS/src/openms/source/APPLICATIONS/TOPPBase.cpp:444
#8  0x00005555555884ff in main (argc=6, argv=0x7fffffffe098) at /home/sachsenb/OMS/OpenMS/src/topp/IDFileConverter.cpp:829
(gdb) frame 3
#3  0x00007ffff5e1c968 in OpenMS::ProteinIdentification::SearchParameters::getChargeRange (this=0x55555596cd20) at /home/sachsenb/OMS/OpenMS/src/openms/source/METADATA/ProteinIdentification.cpp:208
208           result.first = charges.toInt();
(gdb) print charges
$1 = {<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >> = "+1, +2", static EMPTY = {<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >> = "",
    static EMPTY = <same as static member of an already seen type>}}
(gdb) frame 2
#2  0x00007ffff5d74e1a in OpenMS::String::toInt (this=0x55555596cd88) at /home/sachsenb/OMS/OpenMS/src/openms/source/DATASTRUCTURES/String.cpp:293
293         return StringUtils::toInt(*this);
```

timosachsenberg commented:
Similar charge range problem in other test
```
Starting program: /home/sachsenb/OMS/OpenMS/openms-build/src/tests/class_tests/bin/IdentificationDataConverter_test
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
checking (void importIDs(IdentificationData&, const vector<ProteinIdentification>&, const vector<PeptideIdentification>&)) ... Progress of 'converting protein identification runs':

Catchpoint 1 (exception thrown), 0x00007ffff4174d1d in __cxa_throw () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
(gdb) bt
#0  0x00007ffff4174d1d in __cxa_throw () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#1  0x00007ffff5d79fb5 in OpenMS::StringUtils::toInt (this_s=...) at /home/sachsenb/OMS/OpenMS/src/openms/include/OpenMS/DATASTRUCTURES/StringUtils.h:810
#2  0x00007ffff5d74e1a in OpenMS::String::toInt (this=0x5555559b8b28) at /home/sachsenb/OMS/OpenMS/src/openms/source/DATASTRUCTURES/String.cpp:293
#3  0x00007ffff5e1c968 in OpenMS::ProteinIdentification::SearchParameters::getChargeRange (this=0x5555559b8ac0) at /home/sachsenb/OMS/OpenMS/src/openms/source/METADATA/ProteinIdentification.cpp:208
#4  0x00007ffff5ec8d79 in OpenMS::IdentificationDataConverter::importDBSearchParameters_ (pisp=..., id_data=...) at /home/sachsenb/OMS/OpenMS/src/openms/source/METADATA/ID/IdentificationDataConverter.cpp:888
#5  0x00007ffff5ec1956 in OpenMS::IdentificationDataConverter::importIDs (id_data=..., proteins=std::vector of length 2, capacity 2 = {...}, peptides=std::vector of length 3, capacity 4 = {...})
    at /home/sachsenb/OMS/OpenMS/src/openms/source/METADATA/ID/IdentificationDataConverter.cpp:74
#6  0x00005555555ae8d7 in main (argc=1, argv=0x7fffffffe0d8) at /home/sachsenb/OMS/OpenMS/src/tests/class_tests/openms/source/IdentificationDataConverter_test.cpp:76
(gdb) frame 4
#4  0x00007ffff5ec8d79 in OpenMS::IdentificationDataConverter::importDBSearchParameters_ (pisp=..., id_data=...) at /home/sachsenb/OMS/OpenMS/src/openms/source/METADATA/ID/IdentificationDataConverter.cpp:888
888         pair<int, int> charge_range = pisp.getChargeRange();
(gdb) print charge_range
$1 = {first = -18561, second = 32767}
(gdb) frame 5
#5  0x00007ffff5ec1956 in OpenMS::IdentificationDataConverter::importIDs (id_data=..., proteins=std::vector of length 2, capacity 2 = {...}, peptides=std::vector of length 3, capacity 4 = {...})
    at /home/sachsenb/OMS/OpenMS/src/openms/source/METADATA/ID/IdentificationDataConverter.cpp:74
74              importDBSearchParameters_(prot.getSearchParameters(), id_data);
(gdb) frame 3
#3  0x00007ffff5e1c968 in OpenMS::ProteinIdentification::SearchParameters::getChargeRange (this=0x5555559b8ac0) at /home/sachsenb/OMS/OpenMS/src/openms/source/METADATA/ProteinIdentification.cpp:208
208           result.first = charges.toInt();
(gdb) print charges
$2 = {<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >> = "1, 2", static EMPTY = {<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >> = "",
    static EMPTY = <same as static member of an already seen type>}}
```

jpfeuffer commented:
But isn't that caught?

Commit:
small fixes regarding empty fields

Commit:
Merge remote-tracking branch 'upstream/develop' into fix/hendrik_idfile_integration

Commit:
force check of input file name

timosachsenberg committed:
Update IdentificationData.cpp

Commit:
make inputfile mandatory in observation

Commit:
Merge remote-tracking branch 'upstream/fix/hendrik_idfile_integration' into fix/hendrik_idfile_integration

hendrikweisser commented:
@timosachsenberg, @jpfeuffer: As discussed on our call, I've made `Observation::input_file` a required attribute and pushed the changes to my branch.
Due to this, the recent commit https://github.com/hendrikweisser/OpenMS/pull/14/commits/66774fb5b6defb2f40eede1a18d17d602a9b4c78 may have introduced some conflicts. I'd suggest to revert that commit and pull from my "idfile_integration" branch again.

timosachsenberg commented:
Please don't do that in the future.
I also added some other fixes and now we need to find out how to get those back in.

timosachsenberg commented:
I will give it a try but I would be happy if you could take a look at it after that.

hendrikweisser commented:
> Please don't do that in the future.

What? I've made the changes that we agreed I should make.
What do you suggest I should have done differently?

Commit:
Merge remote-tracking branch 'hendrikweisser/idfile-integration' into fix/hendrik_idfile_integration

timosachsenberg committed:
Merge branch 'idfile-integration' into fix/hendrik_idfile_integration

timosachsenberg commented:
No worries. Wasn't that bad. Was probably a misunderstanding from my side.

timosachsenberg committed:
Update IdentificationData.cpp

timosachsenberg commented:
There still seems to be tests failing in ProteomicsLFQ. Maybe still a good idea to get the changes into your branch (e.g., I updated for changes in develop).

timosachsenberg commented:
```
The following tests FAILED:
        2421 - UTILS_ProteomicsLFQ_2 (Failed)
        2422 - UTILS_ProteomicsLFQ_2_out_1 (Failed)
        2424 - UTILS_ProteomicsLFQ_2_out_3 (Failed)
        2427 - UTILS_ProteomicsLFQ_3_out_1 (Failed)
```
